bench/range.hpp

```diff
@@ -0,0 +1,238 @@
+    auto rv1 = r2.finished() ? r1.yield() : r1.peek();
+    return std::make_tuple(rv1, rv2, rv3, rv4);
+  }
+  bool finished() { return r1.finished(); }
+};
+
+template <typename Range1, typename Range2, typename Range3, typename Range4,
+          typename Range5>
+class Range5D
+    : public Range<std::tuple<typename Range1::elem_t, typename Range2::elem_t,
+                              typename Range3::elem_t, typename Range4::elem_t,
+                              typename Range5::elem_t>> {
+  Range1 r1;
+  Range2 r2;
+  Range3 r3;
+  Range4 r4;
+  Range5 r5;
+
+ public:
+  typedef std::tuple<typename Range1::elem_t, typename Range2::elem_t,
+                     typename Range3::elem_t, typename Range4::elem_t,
+                     typename Range5::elem_t>
+      elem_t;
+  Range5D(Range1 _r1, Range2 _r2, Range3 _r3, Range4 _r4, Range5 _r5)
+      : r1(_r1), r2(_r2), r3(_r3), r4(_r4), r5(_r5) {}
+  elem_t peek() {
+    return std::make_tuple(r1.peek(), r2.peek(), r3.peek(), r4.peek(),
+                           r5.peek());
+  }
+  elem_t yield() {
+    auto rv5 = r5.yield();
+    auto rv4 = r5.finished() ? r4.yield() : r4.peek();
+    auto rv3 = r4.finished() ? r3.yield() : r3.peek();
+    auto rv2 = r3.finished() ? r2.yield() : r2.peek();
+    auto rv1 = r2.finished() ? r1.yield() : r1.peek();
+    return std::make_tuple(rv1, rv2, rv3, rv4, rv5);
+  }
+  bool finished() { return r1.finished(); }
+};
+
+/**
+ * Utility range constructors.
+ */
+template <typename T>
+SizeRange<T> size_range(T low, T high, T mult) {
+  return SizeRange<T>(low, high, mult);
+}
+
+template <typename Vect>
+ValueRange<typename Vect::value_type> value_range(Vect vals) {
+  return ValueRange<typename Vect::value_type>(vals);
+}
+
+template <typename T>
+ValueRange<T> value_range(std::initializer_list<T> l) {
+  return ValueRange<T>(l);
+}
+
+template <typename Range1, typename Range2>
+Range2D<Range1, Range2> nd_range(Range1 r1, Range2 r2) {
+  return Range2D<Range1, Range2>(r1, r2);
+}
+
+template <typename Range1, typename Range2, typename Range3>
+Range3D<Range1, Range2, Range3> nd_range(Range1 r1, Range2 r2, Range3 r3) {
+  return Range3D<Range1, Range2, Range3>(r1, r2, r3);
+}
+
+template <typename Range1, typename Range2, typename Range3, typename Range4>
+Range4D<Range1, Range2, Range3, Range4> nd_range(Range1 r1, Range2 r2,
+                                                 Range3 r3, Range4 r4) {
+  return Range4D<Range1, Range2, Range3, Range4>(r1, r2, r3, r4);
+}
+
+template <typename Range1, typename Range2, typename Range3, typename Range4,
+          typename Range5>
+Range5D<Range1, Range2, Range3, Range4, Range5> nd_range(Range1 r1, Range2 r2,
+                                                         Range3 r3, Range4 r4,
+                                                         Range5 r5) {
+  return Range5D<Range1, Range2, Range3, Range4, Range5>(r1, r2, r3, r4, r5);
+}
+
+#endif  // include guard
```

Comment:
is this coming from elsewhere? is there a link to the source?

Reply:
Do you mean the source code in range.hpp? No, it's all written by me...